Mathlib/LinearAlgebra/TensorProduct/Vanishing.lean

```diff
@@ -0,0 +1,256 @@
+It is not difficult to show (`TensorProduct.sum_tmul_eq_zero_of_vanishesTrivially`) that if
+$\sum_i m_i \otimes n_i$ vanishes trivially, then it vanishes; that is,
+$\sum_i m_i \otimes n_i = 0$.
+
+The *equational criterion for vanishing* (`TensorProduct.vanishesTrivially_iff_sum_tmul_eq_zero`),
+which appears as
+[A. Altman and S. Kleiman, *A term of commutative algebra* (Lemma 8.16)][altman2021term],
+states that if the elements $m_i$ generate the module $M$, then $\sum_i m_i \otimes n_i = 0$ if and
+only if the expression $\sum_i m_i \otimes n_i$ vanishes trivially.
+
+We also prove the following generalization
+(`TensorProduct.vanishesTrivially_iff_sum_tmul_eq_zero_of_rTensor_injective`). If the submodule
+$M' \subseteq M$ generated by the $m_i$ satisfies the property that the induced map
+$M' \otimes N \to M \otimes N$ is injective, then $\sum_i m_i \otimes n_i = 0$ if and only if the
+expression $\sum_i m_i \otimes n_i$ vanishes trivially. (In the case that $M = R$, this yields the
+"equational criterion for flatness".)
+
+Conversely (`TensorProduct.rTensor_injective_of_forall_vanishesTrivially`),
+suppose that for every equation $\sum_i m_i \otimes n_i = 0$, the expression
+$\sum_i m_i \otimes n_i$ vanishes trivially. Then the induced map $M' \otimes N \to M \otimes N$
+is injective for every submodule $M' \subseteq M$.
+
+## References
+
+* [A. Altman and S. Kleiman, *A term of commutative algebra* (Lemma 8.16)][altman2021term]
+
+## TODO
+
+* Prove the same theorems with $M$ and $N$ swapped.
+* Prove the same theorems with universe polymorphism.
+
+-/
+
+universe u
+
+variable (R : Type u) [CommRing R]
+variable {M : Type u} [AddCommGroup M] [Module R M]
+variable {N : Type u} [AddCommGroup N] [Module R N]
+
+open Classical DirectSum LinearMap Function Submodule
+open scoped BigOperators
+
+namespace TensorProduct
+
+variable {ι : Type u} [Fintype ι] {m : ι → M} {n : ι → N}
+
+variable (m n) in
+/-- The proposition that the expression $\sum_i m_i \otimes n_i$ in $M \otimes N$
+*vanishes trivially*. That is, there exist a finite index type $\kappa$,
+elements $(y_j)_{j \in \kappa}$ of $N$, and elements $(a_{ij})_{i \in \iota, j \in \kappa}$ of $R$
+such that for all $i$,
+$$n_i = \sum_j a_{ij} y_j$$
+and for all $j$,
+$$\sum_{i} a_{ij} m_i = 0.$$
+Note that this condition is not symmetric in $M$ and $N$. -/
+abbrev VanishesTrivially : Prop :=
+  ∃ (κ : Type u) (_ : Fintype κ) (a : ι → κ → R) (y : κ → N),
+    (∀ i, n i = ∑ j, a i j • y j) ∧ ∀ j, ∑ i, a i j • m i = 0
+
+/-- **Equational criterion for vanishing**
+[A. Altman and S. Kleiman, *A term of commutative algebra* (Lemma 8.16)][altman2021term],
+backward direction.
+
+If the expression $\sum_i m_i \otimes n_i$ vanishes trivially, then it vanishes.
+That is, $\sum_i m_i \otimes n_i = 0$. -/
+theorem sum_tmul_eq_zero_of_vanishesTrivially (hmn : VanishesTrivially R m n) :
+    ∑ i, m i ⊗ₜ n i = (0 : M ⊗[R] N) := by
```

Comment:
Should there be a variant of `VanishesTrivially` with support for `s : Finset ι`, so that you can consider `∑ i in s, m i ⊗ₜ n i`?
That might lead to an API that can be applied more flexibly in applications.

Reply:
I will try to just change the definition of `VanishesTrivially` to work with sums over a `Finset`, because I think it would be confusing to have two versions. This might take a while, though.

Reply:
I can't figure out how to do this without it being very painful. Also, the added flexibility is minimal. In some situations it can make using the results more cumbersome, because now the set `s` must be an argument of `VanishesTrivially`, and now the first condition of `VanishesTrivially` requires the additional hypothesis `i ∈ s`.

Reply:
Ok, thanks for trying!